metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -532,113 +745,107 @@ func join(fields ...string) string {
-	replicasetWatcher kubernetes.Watcher,
-	jobWatcher kubernetes.Watcher,
+	metricsetName string,
+	resourceName string,
+	resourceWatchers *Watchers,
+	config *kubernetesConfig,
 	update func(map[string]mapstr.M, kubernetes.Resource),
 	delete func(map[string]mapstr.M, kubernetes.Resource),
 	index func(e mapstr.M) string) *enricher {
 
 	enricher := enricher{
-		metadata:          map[string]mapstr.M{},
-		index:             index,
-		watcher:           watcher,
-		nodeWatcher:       nodeWatcher,
-		namespaceWatcher:  namespaceWatcher,
-		replicasetWatcher: replicasetWatcher,
-		jobWatcher:        jobWatcher,
-	}
-
-	watcher.AddEventHandler(kubernetes.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			enricher.Lock()
-			defer enricher.Unlock()
-			update(enricher.metadata, obj.(kubernetes.Resource))
-		},
-		UpdateFunc: func(obj interface{}) {
-			enricher.Lock()
-			defer enricher.Unlock()
-			update(enricher.metadata, obj.(kubernetes.Resource))
-		},
-		DeleteFunc: func(obj interface{}) {
-			enricher.Lock()
-			defer enricher.Unlock()
-			delete(enricher.metadata, obj.(kubernetes.Resource))
-		},
-	})
+		metadata:      map[string]mapstr.M{},
+		index:         index,
+		resourceName:  resourceName,
+		metricsetName: metricsetName,
+		config:        config,
+	}
+
+	resourceWatchers.lock.Lock()
+	defer resourceWatchers.lock.Unlock()
+
+	watcher := resourceWatchers.watchersMap[resourceName]
+	if watcher != nil {
+		watcher.watcher.AddEventHandler(kubernetes.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				enricher.Lock()
+				defer enricher.Unlock()
+				update(enricher.metadata, obj.(kubernetes.Resource))
+			},
+			UpdateFunc: func(obj interface{}) {
+				enricher.Lock()
+				defer enricher.Unlock()
+				update(enricher.metadata, obj.(kubernetes.Resource))
+			},
+			DeleteFunc: func(obj interface{}) {
+				enricher.Lock()
+				defer enricher.Unlock()
+				delete(enricher.metadata, obj.(kubernetes.Resource))
+			},
+		})
+	}
 
 	return &enricher
 }
 
-func (m *enricher) Start() {
-	m.watchersStartedLock.Lock()
-	defer m.watchersStartedLock.Unlock()
-	if !m.watchersStarted {
-		if m.nodeWatcher != nil {
-			if err := m.nodeWatcher.Start(); err != nil {
-				logp.Warn("Error starting node watcher: %s", err)
-			}
-		}
-
-		if m.namespaceWatcher != nil {
-			if err := m.namespaceWatcher.Start(); err != nil {
-				logp.Warn("Error starting namespace watcher: %s", err)
-			}
-		}
+func (e *enricher) Start(resourceWatchers *Watchers) {
+	resourceWatchers.lock.Lock()
+	defer resourceWatchers.lock.Unlock()
 
-		if m.replicasetWatcher != nil {
-			if err := m.replicasetWatcher.Start(); err != nil {
-				logp.Warn("Error starting replicaset watcher: %s", err)
-			}
+	resourceWatcher := resourceWatchers.watchersMap[e.resourceName]
+	if resourceWatcher != nil && !resourceWatcher.started {
+		if err := resourceWatcher.watcher.Start(); err != nil {
+			log.Warnf("Error starting %s watcher: %s", e.resourceName, err)
```

Comment:
Why are we using another log library here? Can't we just use the `logp` ?  Same comment for Stop enricher.

Reply:
Removed it. I passed the old logger as parameter to the enricher now. `logp` is being used for everything currently